### PR TITLE
CA-184556: improve efficiency handling many VBDs during VM lifecycle operations

### PR DESCRIPTION
**Pull request by mg12**

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
A few comments here. Think probably best we have a sync-up and talk through some of it. Maybe some of the concerns are unfounded but we need to check.

**Comment by mg12** [comment]
I have updated the patches with all the comments from the code review, and split it into two parts, one for the cache and the other for the Parallel atom. This pull request is for the cache only, I'll create a new one for the second part.

File: xc/xenops_server_xen.ml [hunk 1]
```diff
@@ -448,24 +448,87 @@ end
...
+*)
+
+module DeviceCache = struct
+	module PerVMCache = struct
+		include Hashtbl
+		let create n = (create n, Mutex.create ())
+	end
+	include Hashtbl
+	let create n = (create n, Mutex.create())
+	let discard (cache, mutex) domid = Mutex.execute mutex (fun () ->
```
Comment (jonludlam) [inline xc/xenops_server_xen.ml]: Looks like there's some duplication going on here (2 include Hashtbls, 2 'let create' functions)

Comment (mg12) [inline xc/xenops_server_xen.ml]: there are 2 levels of hashtables:
- the lowest level is a hashtable that keeps the cached devices for a specific domid
- the highest level is a hashtable the keeps the lowest level hashtables

The create for each of the levels has been redefined to return not only what create would return, but also a unique mutex for that hashtable instance.

**Comment by simonjbeaumont** [comment]
OK, I think we've converged on this one.

**Comment by simonjbeaumont** [comment]
Adding blocked label until review of #250 is finished

**Pull request merged**